### PR TITLE
feat(ui5-li-notification, ui5-li-notification-group): introduce new components

### PR DESCRIPTION
**Pull request by ilhan007**
The ui5-li-notification and ui5-li-notification-group components  cover the main functionality of their equivalents in openui5 - sap.m.NotificationListItem and sap.m.NotificationGroup. They are meant to display  app notifications and they possess a rich set of properties and slots to achieve it.

Fixes: https://github.com/SAP/ui5-webcomponents/issues/1478

Screenshot 2020-05-07 at 7 46 52 AM

### CODE REVIEW

File: packages/fiori/src/types/Priority.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import DataType from "@ui5/webcomponents-base/dist/types/DataType.js";
+
+/**
+ * Different types of Priority.
+ * @lends sap.ui.webcomponents.main.types.Priority.prototype
```
[the remark this answers is not in the document]
Comment (vladitasev) [reply packages/fiori/src/types/Priority.js:5]: Now that I think of it, is it possible to need this type for some "main" component in the future? Maybe move it to "main" entirely. I'll do some research.

File: packages/fiori/src/NotificationOverflowAction.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		 * @public
+		 */
+		icon: {
+			type: String,
+		},
+	},
+	slots: {
+	},
+	events: {
+		click: {},
```
[the remark this answers is not in the document]
Comment (MarcusNotheis) [reply packages/fiori/src/NotificationOverflowAction.js:37]: It depends - in order to make them work: no. If you want them to show in in your playground/ our storybook you to add it (same as the `ui5-button` click). I would prefer to have them added to be honest.

File: packages/fiori/src/NotificationListItem.js [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+	_onclick(event) {
+		if (event.isMarked === "button") {
+			return;
+		}
+		this.fireItemPress(event);
+	}
+
+	_onShowMoreClick() {
+		this._showMorePressed = !this._showMorePressed;
+		this.noTruncation = !this.noTruncation;
```
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/fiori/src/NotificationListItem.js:400]: noTruncation is the private property, the effective state of the truncation (it is confiusing having the public "truncate")

File: packages/fiori/src/NotificationListItem.js [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+
+	get standardActions() {
+		if (this.actions.length > 1) {
+			return [];
+		}
+
+		return this.clonedActions;
+	}
+
+	get clonedActions() {
```
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/fiori/src/NotificationListItem.js:367]: agree

File: packages/fiori/src/NotificationListItem.js [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+		 *
+		 * <b>Note:</b> Consider using the <code>ui5-notification-action</code>.
+		 *
+		 * @type {HTMLElement}
+		 * @slot
+		 * @public
+		 */
+		actions: {
+			type: HTMLElement,
+			propertyName: "actions",
```
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/fiori/src/NotificationListItem.js:115]: yes

File: packages/fiori/src/NotificationListItem.js [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+import ResizeHandler from "@ui5/webcomponents-base/dist/delegate/ResizeHandler.js";
+import { isIE } from "@ui5/webcomponents-base/dist/Device.js";
+
+import ListItemBase from "@ui5/webcomponents/dist/ListItemBase.js";
+import Button from "@ui5/webcomponents/dist/Button.js";
+import Icon from "@ui5/webcomponents/dist/Icon.js";
+import Priority from "./types/Priority.js";
+
+// icons
+import "@ui5/webcomponents-icons/dist/icons/decline.js";
```
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/fiori/src/NotificationListItem.js:13]: added

File: packages/fiori/src/NotificationListItem.js [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+
+		/**
+		 * Defines if the <code>heading</code> and <code>decription</code> should truncate,
+		 * they would wrap by default.
+		 * @type {boolean}
+		 * @defaultvalue false
+		 * @public
+		 */
+		truncate: {
+			type: Boolean,
```
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/fiori/src/NotificationListItem.js:83]: agree, I will do some renaming

File: packages/fiori/src/NotificationOverflowAction.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/fiori/src/NotificationOverflowAction.js:37]: As Marcus said, we should describe it in order to have it in the API reference.

File: packages/fiori/src/types/Priority.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/fiori/src/types/Priority.js:5]: I added it first in main, then changed my mind and moved to fiori. It sounds general but we should check if it is at lease used somewhere else I guess

File: packages/main/src/themes/List.css [hunk 8]
```diff
@@ -1,6 +1,7 @@
 :host(:not([hidden])) {
 	display: block;
 	max-width: 100%;
+	width: 100%;
```
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/main/src/themes/List.css:4]: I will document it, you are right It might visually breaking for applications, I prefer it 100% wide as well. The native "ul" element is also 100% wide.

File: packages/fiori/src/NotificationListItem.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (ilhan007) [reply packages/fiori/src/NotificationListItem.js:85]: done

File: packages/fiori/src/NotificationListItemBase.js [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+			defaultValue: Priority.None,
+		},
+
+		/**
+		 * Defines if the <code>close</code> button would be displayed.
+		 * @type {boolean}
+		 * @defaultvalue false
+		 * @public
+		 */
+		showClose: {
```
Comment (fifoosid) [inline packages/fiori/src/NotificationListItemBase.js:61]: I think it would be better if by default the close button is present, so the property is something like hideClose

File: packages/fiori/src/NotificationListItemBase.js [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+ * @constructor
+ * @author SAP SE
+ * @alias sap.ui.webcomponents.fiori.NotificationListItemBase
+ * @extends ListItemBase
+ * @tagname ui5-li-notification-group
+ * @since 1.0.0-rc.8
+ * @appenddocs NotificationOverflowAction
+ * @public
+ */
+class NotificationListItemBase extends ListItemBase {
```
Comment (fifoosid) [inline packages/fiori/src/NotificationListItemBase.js:106]: Should import ```ui5-popover``` and await it to be defined.

File: packages/fiori/src/NotificationListGroupItem.hbs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+	tabindex="{{_tabIndex}}"
+	dir="{{rtl}}"
+	aria-labelledby="{{ariaLabelledBy}}"
+>
+	<div class="ui5-nli-group-header">
+		<ui5-button
+			icon="navigation-right-arrow"
+			design="Transparent"
+			@click="{{_onBtnToggleClick}}"
+			class="ui5-nli-group-toggle-btn"
```
Comment (fifoosid) [inline packages/fiori/src/NotificationListGroupItem.hbs:16]: I think that you only need either the class or the toggle-btn attribute. As much as I can see you are using both of them to identify this exact DOM node. Moreover you are using the same attribute in ```NotificationListItetm.hbs```. Probably you could go with the class only?

File: packages/fiori/src/NotificationListGroupItem.hbs [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+		{{/if}}
+
+		<div class="ui5-nli-group-divider"></div>
+
+		{{#unless collapsed}}
+			{{#if showOverflow}}
+				<ui5-button
+					icon="overflow"
+					design="Transparent"
+					@click="{{_onBtnOverflowClick}}"
```
Comment (fifoosid) [inline packages/fiori/src/NotificationListGroupItem.hbs:42]: Same here

File: packages/fiori/src/NotificationListGroupItem.hbs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+					class="ui5-nli-group-overflow-btn"
+					title="{{overflowBtnTitle}}"
+					overflow-btn
+				></ui5-button>
+			{{else}}
+				{{#each standardActions}}
+					<ui5-button
+						icon="{{this.icon}}"
+						class="ui5-nli-group-action"
```
Comment (fifoosid) [inline packages/fiori/src/NotificationListGroupItem.hbs:51]: Same here

File: packages/fiori/src/NotificationListItem.hbs [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+	dir="{{rtl}}"
+	aria-labelledby="{{ariaLabelledBy}}"
+>
+	<div class="ui5-nli-actions">
+		{{#if showOverflow}}
+			<ui5-button
+				icon="overflow"
+				design="Transparent"
+				@click="{{_onBtnOverflowClick}}"
+				class="ui5-nli-overflow-btn"
```
Comment (fifoosid) [inline packages/fiori/src/NotificationListItem.hbs:19]: Same here

File: packages/fiori/src/NotificationListItem.hbs [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+				title="{{overflowBtnTitle}}"
+				overflow-btn
+			></ui5-button>
+		{{else}}
+			{{#each standardActions}}
+				<ui5-button
+					icon="{{this.icon}}"
+					class="ui5-nli-action"
```
Comment (fifoosid) [inline packages/fiori/src/NotificationListItem.hbs:27]: Same here

File: packages/fiori/src/NotificationListItemBase.js [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+import "@ui5/webcomponents-icons/dist/icons/message-warning.js";
+import "@ui5/webcomponents-icons/dist/icons/overflow.js";
+
+// Templates
+import NotifactionOverflowActionsPopoverTemplate from "./generated/templates/NotifactionOverflowActionsPopoverTemplate.lit.js";
+
+// Styles
+import NotifactionOverflowActionsPopoverCss from "./generated/themes/NotifactionOverflowActionsPopover.css.js";
+
+const PRIORITY_ICONS_MAP = {
```
Comment (fifoosid) [inline packages/fiori/src/NotificationListItemBase.js:21]: This could be a static method.

File: packages/fiori/src/NotificationListItem.js [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+
+		if (this.priority === Priority.Low) {
+			return this.i18nBundle.getText(NOTIFICATION_LIST_ITEM_LOW_PRIORITY_TXT);
+		}
+
+		return "";
+	}
+
+	get accInvisibleText() {
+		const notifcatationTxt = this.i18nBundle.getText(NOTIFICATION_LIST_ITEM_TXT);
```
Comment (vladitasev) [inline packages/fiori/src/NotificationListItem.js:343]: typo

File: packages/fiori/src/NotificationListGroupItem.js [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+ * @extends NotificationListItemBase
+ * @tagname ui5-li-notification-group
+ * @since 1.0.0-rc.8
+ * @appenddocs NotificationOverflowAction
+ * @public
+ */
+class NotificationListGroupItem extends NotificationListItemBase {
+	constructor() {
+		super();
+		this.i18nBundle = getI18nBundle("@ui5/webcomponents-fiori");
```
Comment (vladitasev) [inline packages/fiori/src/NotificationListGroupItem.js:118]: Maybe this can go to the parent constructor and remove this constructor altogether? Not so important though

File: packages/fiori/src/NotificationListGroupItem.js [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+			fetchI18nBundle("@ui5/webcomponents-fiori"),
+		]);
+	}
+
+	get itemsCount() {
+		return this.items.length;
+	}
+
+	get counter() {
+		return `(${this.itemsCount})`;
```
Comment (vladitasev) [inline packages/fiori/src/NotificationListGroupItem.js:151]: I think it would be simpler to move the `(` and `)` to the template and use `itemsCount` directly and remove this getter

File: docs/Public Module Imports.md [hunk 20]
```diff
@@ -136,6 +136,8 @@ For API documentation and samples, please check the [UI5 Web Components Playgrou
 | Shell Bar Item           | `ui5-shellbar-item`  | `import "@ui5/webcomponents-fiori/dist/ShellBarItem.js";`  |
 | Product Switch           | `ui5-product-switch`       | `import "@ui5/webcomponents-fiori/dist/ProductSwitch.js";`      |
 | Product Switch Item      | `ui5-product-switch-item`  | `import "@ui5/webcomponents-fiori/dist/ProductSwitchItem.js";`  |
+| Notification List Item      | `ui5-li-notifcation`      | `import "@ui5/webcomponents-fiori/dist/NotifcationListItem.js";`     |
```
Comment (vladitasev) [inline docs/Public Module Imports.md:139]: Also add `NotificationOverflowAction`

File: packages/fiori/src/NotificationListItem.js [hunk 21]
```diff
@@ -0,0 +1,432 @@
...
+		const readTxt = this.read ? this.i18nBundle.getText(NOTIFICATION_LIST_ITEM_READ) : this.i18nBundle.getText(NOTIFICATION_LIST_ITEM_UNREAD);
+		const priorityText = this.priorityText;
+
+		return `${notifcatationTxt} ${readTxt} ${priorityText}`;
```
Comment (vladitasev) [inline packages/fiori/src/NotificationListItem.js:347]: same

File: packages/fiori/src/UploadCollectionItem.js [hunk 22]
```diff
@@ -257,13 +257,13 @@ class UploadCollectionItem extends ListItem {
 	}
 
 	onBeforeRendering() {
-		if (!this.focused) {
```
Comment (vladitasev) [inline packages/fiori/src/UploadCollectionItem.js:260]: Is this file intended to be changed?

File: packages/fiori/src/themes/InvisibleTextStyles.css [hunk 23]
```diff
@@ -0,0 +1,7 @@
+.ui5-hidden-text {
```
Comment (vladitasev) [inline packages/fiori/src/themes/InvisibleTextStyles.css:1]: Not for this change, but we should consider moving this somewhere else - base or theme-base

File: packages/main/src/List.js [hunk 24]
```diff
@@ -179,6 +179,37 @@ const metadata = {
 			},
 		},
 
+		/**
```
Comment (vladitasev) [inline packages/main/src/List.js:182]: Have you considered having a new list? Or it's an overkill? I'm fine with this solution, this will be very minimal when code is minified.

Comment (ilhan007) [reply packages/main/src/List.js:182]: Yes I thought about this, as these events are specific to notifications, at the end I decided that it not so much code in addition.

File: packages/fiori/src/themes/InvisibleTextStyles.css (hunk 23 above)
Comment (ilhan007) [reply packages/fiori/src/themes/InvisibleTextStyles.css:1]: I agree

File: packages/fiori/src/UploadCollectionItem.js (hunk 22 above)
Comment (ilhan007) [reply packages/fiori/src/UploadCollectionItem.js:260]: > Is this file intended to be changed?

It is intended. This was the reason the PR was failing. From one hand we dont want the entire lis item to get the outline when we interact with ui5-button ui5-link and I fixed this, but the UploadCollection used to depend this.focused to turn on/off the editing mode, now it relies on its own this._focused.

File: docs/Public Module Imports.md (hunk 20 above)
Comment (ilhan007) [reply docs/Public Module Imports.md:139]: ok

File: packages/fiori/src/NotificationListGroupItem.js (hunk 18 above)
Comment (ilhan007) [reply packages/fiori/src/NotificationListGroupItem.js:118]: ok

File: packages/fiori/src/NotificationListGroupItem.js (hunk 19 above)
Comment (ilhan007) [reply packages/fiori/src/NotificationListGroupItem.js:151]: Done

File: packages/fiori/src/NotificationListItemBase.js (hunk 10 above)
Comment (ilhan007) [reply packages/fiori/src/NotificationListItemBase.js:106]: done

**Comment by vladitasev** [comment]
Please also resolve conflicts
